docs/versioning.md

```diff
@@ -0,0 +1,567 @@
+# Versioning
+
+This document describes S3 Server's support for the specific AWS S3 Bucket
+Versioning feature.
+
+## AWS S3 Bucket Versioning
+
+See AWS documentation on Versioning for a description of the Bucket Versioning
+feature:
+
+http://docs.aws.amazon.com/AmazonS3/latest/dev/Versioning.html
+http://docs.aws.amazon.com/AmazonS3/latest/dev/ObjectVersioning.html
+
+## Implementation of Bucket Versioning in S3
+
+### Overview of Metadata and API Component Roles
+
+Each version of an object is stored as a separate key in metadata. The S3 API
+interacts with the metadata backend to store, retrieve, and delete version
+metadata.
+
+The implementation of versioning within the metadata backend is naive. The
+metadata backend does not evaluate any information about bucket or version state
+(whether versioning is enabled or suspended, and whether a version is a null
+version or delete marker). The S3 front-end API manages the logic regarding
+versioning information, and sends instructions to metadata to handle the basic
+CRUD operations for version metadata.
+
+The role of the S3 API can be broken down into the following:
+
+- put and delete version data
+- store extra information about a version, such as whether it is a delete marker
+  or null version, in the object's metadata
+- send instructions to metadata backend to store, retrieve, update and delete
+  version metadata based on bucket versioning state and version metadata
+- encode version ID information to return in responses to requests, and decode
+  version IDs sent in requests
+
+The implementation of Bucket Versioning in S3 is described in this document in
+two main parts. The first section, "Implementation of Bucket Versioning in
+Metadata" describes the way versions are stored in metadata, and the metadata
+options for manipulating version metadata.
+
+The second section, "Implementation of Bucket Versioning in API" describes the
+way the metadata options are used in the API within S3 actions to create new
+versions, update their metadata, and delete them. The management of null
+versions and creation of delete markers are also described in this section.
+
+## Implementation of Bucket Versioning in Metadata
+
+As mentioned above, each version of an object is stored as a separate key in
+metadata. We use version identifiers as the suffix for the keys of the object
+versions, and a special version (the "Master Version") to represent the latest
```

Comment:
I added a link so it's clearer that the quotes are referencing a specific section.